tests/php/src/DependencyInjectedTestCase.php

```diff
@@ -0,0 +1,57 @@
+<?php
+
+namespace AmpProject\AmpWP\Tests;
+
+use AmpProject\AmpWP\AmpWpPlugin;
+use AmpProject\AmpWP\Infrastructure\Injector;
+use AmpProject\AmpWP\Infrastructure\ServiceContainer;
+use AmpProject\AmpWP\Services;
+use AmpProject\AmpWP\Tests\Helpers\PrivateAccess;
+use WP_UnitTestCase;
+
+abstract class DependencyInjectedTestCase extends WP_UnitTestCase {
+
+	use PrivateAccess;
+
+	/**
+	 * Plugin instance to test with.
+	 *
+	 * @var AmpWpPlugin
+	 */
+	protected $plugin;
+
+	/**
+	 * Service container instance to test with.
+	 *
+	 * @var ServiceContainer
+	 */
+	protected $container;
+
+	/**
+	 * Injector instance to test with.
+	 *
+	 * @var Injector
+	 */
+	protected $injector;
+
+	/**
+	 * Runs the routine before each test is executed.
+	 */
+	public function setUp() {
+		parent::setUp();
+
+		// We're intentionally avoiding the AmpWpPluginFactory here as it uses a
+		// static instance, because its whole point is to allow reuse across consumers.
+		$this->plugin = new AmpWpPlugin();
+		$this->plugin->register();
+
+		$this->container = $this->plugin->get_container();
+		$this->injector  = $this->container->get( 'injector' );
+
+		// The static Services helper has to be modified to use the same objects
+		// as the ones that are injected into the tests.
+		$this->set_private_property( Services::class, 'plugin', $this->plugin );
+		$this->set_private_property( Services::class, 'container', $this->container );
+		$this->set_private_property( Services::class, 'injector', $this->injector );
```

Comment:
Should not these be reset on `tearDown`?

Reply:
We could unset them, yes. I think that shouldn't have an impact, though. If it has, I'd assume we're doing something wrong again in how we test.

Reply:
Looks good: 589e8bc9a3311ce2014f419aac426e9925c62eec

Reply:
Pushed another commit as the comment didn't make sense in the teardown.